internal/server/boltdbstate/project.go

```diff
@@ -69,10 +69,185 @@ func (s *State) ProjectGet(ref *pb.Ref_Project) (*pb.Project, error) {
 // delete. This will delete all operations associated with this project
 // as well.
 func (s *State) ProjectDelete(ref *pb.Ref_Project) error {
+	// Get our project to delete
+	project, err := s.ProjectGet(ref)
+	if err != nil {
+		if status.Code(err) == codes.NotFound {
+			return nil
+		}
+		return err
+	}
+
+	// We perform all of our reads before our write to avoid the deadlocked state
```

Comment:
This seems lossy; we don't lock any of these tables or operations on projects, so in theory operations could be recorded for a project _after_ we preform these reads. 

Examples I can think of would be app polling, and if say someone was in the middle of running a `build` operation while this project delete operation happened.

Reply:
@catsby I agree that we could lose the deletion of records which are added after the deletion of a project begins. After having researched this topic, I learned that having transactions depend upon each other, like we're doing here by listing all of our transactions to delete before we delete them, doesn't work with Bolt's architecture (see [here](https://pkg.go.dev/go.etcd.io/bbolt@v1.3.6#readme-transactions) and [here](https://github.com/boltdb/bolt/issues/240#issuecomment-52409844).

With that said, I'm not sure how else we could retrieve the IDs of all of the records we need to delete during the deletion of a project _without_ first listing them, and thus the write transactions depending upon the reads. I previously considered making the calls to the state package from the  `runner` package. After discussing where to perform the necessary deletions for project-related records, we decided that in the `boltdbstate` package (specifically in the `ProjectDelete` function) would be best, but that was before digging in this far. It might be worth revisiting the idea of calling them from the runner, but then we're back to how to lock specific tables (or "buckets" in Bolt), which I don't think can be done in Bolt. 🤔 

cc/ @cicoyle @briancain

Reply:
@paladin-devops - You want to do these listings inside the `db.Update()` func, which _does_ act as a lock for writes. That also means you'll want to use the lower case state versions, which are meant to be called inside a transaction.

Then in the case @catsby mentioned, it wouldn't be possible to do a write during this operation because this transaction holds onto the table lock.

Reply:
@briancain thanks for the tip!! If I recall, when I was initially attempting to perform the reads and writes from `ProjectDelete`, I was attempting to use the unexported (lowercased) state functions. However, upon looking at that once again now, I've noticed that for `appOperation`, the delete functions don't have unexported versions of themselves - i.e., when they are called, new transactions are unavoidably started (the transaction isn't being passed in to the function, it's being set inside the function):
```go
// Delete deletes an operation record by reference
func (op *appOperation) Delete(s *State, ref *pb.Ref_Operation) error {
	memTxn := s.inmem.Txn(true)
	defer memTxn.Abort()

	target, ok := ref.Target.(*pb.Ref_Operation_Id)
	if !ok {
		return status.Errorf(codes.FailedPrecondition,
			"unknown operation reference type: %T", ref.Target)
	}
	id := target.Id

	err := s.db.Update(func(dbTxn *bolt.Tx) error {
		return op.dbDelete(dbTxn, []byte(id))
	})
	if err == nil {
		memTxn.Commit()
	}

	return err
}
```
Unlike one of the other functions such as `triggerDelete`:
```go
func (s *State) triggerDelete(
	dbTxn *bolt.Tx,
	memTxn *memdb.Txn,
	ref *pb.Ref_Trigger,
) error {
	// Get the trigger. If it doesn't exist then we're successful.
	_, err := s.triggerGet(dbTxn, memTxn, ref)
	if err != nil {
		if status.Code(err) == codes.NotFound {
			return nil
		}

		return err
	}

	// Delete from bolt
	id := s.triggerIdByRef(ref)
	if err := dbTxn.Bucket(triggerBucket).Delete(id); err != nil {
		return err
	}

	// Delete from memdb
	if err := memTxn.Delete(triggerIndexTableName, &triggerIndexRecord{Id: string(id)}); err != nil {
		return err
	}

	return nil
}
```

To address this, I'm considering adding unexported versions of the delete endpoints for each app operation, with additional input params to allow the Bolt and Mem DB transactions to be passed in to the functions, rather than being created during the `Delete` function, leading to the deadlocked state.

Reply:
Yep, you'll need to add an app operation version for a `delete` that can be called in a transaction, makes sense! 👍🏻

Reply:
Updated! 😎 Thank you @catsby & @briancain !